Library/OcCryptoLib/X64/Sha512Avx.nasm

```diff
@@ -0,0 +1,511 @@
+  RORQ    tmp0, 6  ; 34     ; tmp = ((a ror 5) ^ a) ror 6
+  xor     tmp0, a_64        ; tmp = (((a ror 5) ^ a) ror 6) ^ a
+  lea     h_64, [T1 + T2]   ; a(next_state) = T1 + Maj(a,b,c)
+  RORQ    tmp0, 28  ; 28    ; tmp = ((((a ror5)^a)ror6)^a)ror28 = S0(a)
+  add     h_64, tmp0        ; a(next_state) = T1 + Maj(a,b,c) S0(a)
+  RotateState
+%endmacro
+
+; Compute message schedules t and t+1
+%macro SHA512_2Sched 1
+  ; x ROTR n = (x >> n) | (x << 64-n)
+
+  ; sigma[1,512](W[t-2]) = (W[t-2] ROTR 19) ^ (W[t-2] ROTR 61) ^ (W[t-2] SHR 6)
+  %assign idx  (%1 - 2)
+  ; W[t-2] ROTR 19
+  vmovdqu xmm4, W_t(idx)      ; xmm4 = W[t-2]|W[t-1]
+  vpsrlq  xmm0, xmm4, 19      ; xmm0 = W[t-2] >> 19
+  vpsllq  xmm1, xmm4, (64-19) ; xmm1 = W[t-2] << 64-19
+  vpor    xmm0, xmm0, xmm1    ; xmm0 = (W[t-2] >> 19) | (W[t-2] << 64-19)
+  ; W[t-2] ROTR 61
+  vpsrlq  xmm2, xmm4, 61      ; xmm2 = W[t-2] >> 61
+  vpsllq  xmm1, xmm4, (64-61) ; xmm1 = W[t-2] << 64-19
+  vpor   xmm2, xmm2, xmm1     ; xmm2 = (W[t-2] >> 61) | (W[t-2] << 64-61)
+  vpxor	 xmm0, xmm0, xmm2     ; xmm0 = (W[t-2] ROTR 19) ^ (W[t-2] ROTR 61)
+  ; W[t-2] SHR 6
+  vpsrlq  xmm2, xmm4, 6       ; xmm2 = W[t-2] >> 6
+  vpxor	 xmm0, xmm0, xmm2     ; xmm0 = (W[t-2] ROTR 19) ^ (W[t-2] ROTR 61) ^ (W[t-2] SHR 6)
+
+  ; sigma[0,512](W[t-15]) = (W[t-15] ROTR 1) ^ (W[t-15] ROTR 8) ^ (W[t-15] SHR 7)
+  %assign idx  (%1 - 15)
+  ; W[t-15] ROTR 1
+  vmovdqu xmm4, W_t(idx)      ; xmm4 = W[t-15]|W[t-14]
+  vpsrlq  xmm1, xmm4, 1       ; xmm1 = W[t-15] >> 1
+  vpsllq  xmm2, xmm4, (64-1)  ; xmm2 = W[t-15] << 64-1
+  vpor   xmm1, xmm1, xmm2     ; xmm1 = (W[t-15] >> 1) | (W[t-15] << 64-1)
+  ; W[t-15] ROTR 8
+  vpsrlq  xmm3, xmm4, 8       ; xmm3 = W[t-15] >> 8
+  vpsllq  xmm2, xmm4, (64-8)  ; xmm2 = W[t-15] << 64-8
+  vpor   xmm3, xmm3, xmm2     ; xmm3 = (W[t-15] >> 8) | (W[t-15] << 64-8)
+  vpxor	 xmm1, xmm1, xmm3     ; xmm1 = (W[t-15] ROTR 1) ^ (W[t-15] ROTR 8)
+  ; W[t-15] SHR 7
+  vpsrlq  xmm3, xmm4, 7       ; xmm3 = W[t-15] >> 7
+  vpxor	 xmm1, xmm1, xmm3     ; xmm1 = (W[t-15] ROTR 1) ^ (W[t-15] ROTR 8) ^ (W[t-15] SHR 7)
+
+  ; W[t] = sigma[1,512](W[t-2]) + W[t-7] + sigma[0,512](W[t-15]) + W[t-16]
+  %assign idx  (%1 - 7)
+  vpaddq xmm0, xmm0, W_t(idx) ; xmm0 = sigma[1,512](W[t-2]) + W[t-7]
+  vpaddq xmm0, xmm0, xmm1     ; xmm0 = sigma[1,512](W[t-2]) + W[t-7] + sigma[0,512](W[t-15])
+  %assign idx  (%1 - 16)
+  vpaddq xmm0, xmm0, W_t(idx) ; xmm0 = sigma[1,512](W[t-2]) + W[t-7] + sigma[0,512](W[t-15]) + W[t-16]
+  %assign idx  %1
+  vmovdqa W_t(idx), xmm0	    ; Store W[t]
+  vpaddq xmm0, xmm0, K_t(idx) ; Compute W[t]+K[t]
+  vmovdqa WK_2(idx), xmm0     ; Store W[t]+K[t] for next rounds
+%endmacro
+
+; #######################################################################
+; BOOLEAN TryEnableAvx ()
+; To run in QEMU use options: -enable-kvm -cpu Penryn,+avx,+xsave,+xsaveopt
+; #######################################################################
+align 8
+global ASM_PFX(TryEnableAvx)
```

Comment:
What's the plan for calling this to enable new version other than manually, as now? On config.plist setting presumably.

cc @vit9696

Reply:
Yes, a config.plist setting: UEFI → Quirks → EnableVectorAcceleration. Separate from the change in my opinion.